lsp-mode.el

```diff
@@ -5203,18 +5239,44 @@ It will show up only if current point has signature help."
         :context `( :diagnostics ,(lsp-cur-line-diagnostics)
                     ,@(when kind (list :only (vector kind))))))
 
-(defun lsp-code-actions-at-point (&optional kind)
-  "Retrieve the code actions for the active region or the current line."
-  (lsp-request "textDocument/codeAction" (lsp--text-document-code-action-params kind)))
+(defun lsp-code-actions-at-point (&optional kind enabled)
+  "Retrieve the code actions for the active region or the current line.
+If ENABLED is specified, only return code actions that are not
+disabled."
+  (--doto (lsp-request "textDocument/codeAction" (lsp--text-document-code-action-params kind))
+    (when enabled
+      (cl-callf2 seq-remove #'lsp-code-action-disabled? it))))
 
 (defun lsp-execute-code-action-by-kind (command-kind)
   "Execute code action by name."
-  (if-let ((action (->> (lsp-get-or-calculate-code-actions command-kind)
-                        (-filter (-lambda ((&CodeAction :kind?))
-                                   (and kind? (equal command-kind kind?))))
-                        lsp--select-action)))
-      (lsp-execute-code-action action)
-    (signal 'lsp-no-code-actions '(command-kind))))
+  (->> (lsp-get-or-calculate-code-actions command-kind)
+       (-filter (-lambda ((&CodeAction :kind?))
+                  (and kind? (equal command-kind kind?))))
+       ;; Fail in `lsp-execute-code-action' if there is only one disabled
+       ;; action, and show a list of applicable actions with an overview of why
+       ;; they are disabled otherwise.
+       lsp--select-action
+       lsp-execute-code-action))
+
+(defun lsp-execute-code-action-by-type (kind)
```

Comment:
can you elaborate on why we need this method?

Reply:
If there is some specific code action to be executed, e.g. `source.organizeImports`, `lsp-execute-code-action-by-kind` is enough. However, sometimes it can be useful to bring up all quickfixes, or all refactorings: in this case, `lsp-execute-code-action-by-kind` doesn't work, because it must match the kind exactly. My new function matches a base kind: `lsp-execute-code-action-by-type` "refactor" would match "refactor.move", "refactor.*", ... (tested with `jdtls`). This corresponds to the spec's "if the user requests a more specific type of code action, e.g. refactoring".